docs/okeps/okep-5224-connecting-udns/okep-5224-connecting-udns.md

```diff
@@ -0,0 +1,653 @@
+* A new connect-subnet-manager (TBD maybe a controller really?) running only on ovnkube-control-plane
+  i.e cluster-manager pods that will do:
+  * The per node `connectSubnet` slice allocation according to the set
+    `connectSubnet.hostLength` field. A slice of the connect subnet will be handed
+    out to each node and annotated on the node using a new annotation
+    `k8s.ovn.org/node-connect-udn-subnets` that stores the value for each connect
+    UDN API for that node. This mechanism will be very similar to the
+    `k8s.ovn.org/node-subnets` annotation behaviour we have today.
+  * Per node-network router tunnel-key allocation that the topology design
+    (see further sections) needs since there is a requirement of no
+    conflicts of tunnel-keys across nodes. Each connect API would need N*M tunnel keys
+    if N is the number of nodes and M is the number of networks to be connected together.
+    (TBD flesh out annotation data storage ideas...based on which implementation Idea we finalize on (Idea1 v/s Idea2))
+  * TBD: status controller changes
+* A new controller added to OVN-Kubernetes called `ConnectUDNsController` running
+  on each ovnkube-controller that watches the `UserDefinedNetworkAdminConnect` resource will do:
+  * The necessary Northbound database plumbing to provide connectivity between the UDNs.
+  * Read the `k8s.ovn.org/node-connect-udn-subnets` annotation and hand out an IP
+    to each network’s router during interconnection.
+
+NOTE: Having the `NodeNetwork` CRD as an internal data storage API might be more ideal
+to store the above information internally as this needs to be persisted into etcd on reboots.
+It will simplify the annotations and also help with scale since there are known issues
+around patching of annotation and waiting for reading and unmarshalling the annotation
+impacts scale.
+
+#### Network Topology
+
+This section looks at the topology changes that will be introduced to
+connect these three networks together.
+
+**Connecting Layer3 type networks together**
+
+The below diagram shows the overlay part of the OVN Topology that OVN-Kubernetes
+creates for 3 UDN networks (blue, green and yellow) across two nodes.
+
+![3-isolated-l3-networks](https://raw.githubusercontent.com/tssurya/ovn-kubernetes/connecting-udns-okep/docs/okeps/okep-5224-connecting-udns/images/l3-connecting-udns-0.png)
+
+Let’s take a look at how we can connect a blue network, green network and yellow network at the OVN layer.
+Currently there are two ideas that I am torn between, I will cover both
+and outline pros and cons that might make it easier to pick the best one
+during reviews to then dive deeper and add more information. There were other ideas
+which were discarded and listed in the alternatives section.
+
+**Idea1: Add a new transit switch/router between the two ovn_cluster_routers of the networks**
```

Comment:
Do we change Idea 1 to what was Idea 1.1 and just ignore the old way?

Reply:
and the comparison table needs an update? also we probably need to specify routes on the connect_router and UDN routers?